### PR TITLE
Allow employee to choose their own avatar

### LINKED ISSUE #15671: Allow employee to choose their own avatar
**Is your feature request related to a problem? Please describe.**
Since employee cannot choose a theme variation, it becomes hard to know which employee is logged. The only one distinctive element is employee avatar.
employee-top

Currently employee avatar still rely on third party service linked to PrestaShop Forum account. It's not possible to choose an avatar until create an account with same email on PrestaShop Forum.
So it's harder to quickly know who is currently logged as nobody goes to PrestaShop.com to create an account for upload their own avatar.

employee-avatar

**Describe the solution you'd like**
Remove usage of profile.prestashop.com and add an upload field to send a custom image and add a way for module developers to replace the way to retrieve employee avatar by their own implementation.

**Describe alternatives you've considered**
Keep usage of profile.prestashop.com as default implementation but allow module developers to replace it by their own implementation easier.
Currently it require some override to Employee ObjectModel to change behavior of function getImage()
https://github.com/PrestaShop/PrestaShop/blob/10825700128e838a87d91f524d6c529c56a816a4/classes/Employee.php#L609

**Additional context**
On a PrestaShop instance with multiple employee who share same computer, they must open dropdown of employee avatar to see who is currently logged. (Or use different browser...)
By default there are no way to difference who is logged as there are anymore theme variation by employee and process to use a custom avatar is weird.
employee-dropdown

### LINKED ISSUE #18653: Allow users to better handle avatars
### Is your feature request related to a problem?

Currently, as a PrestaShop user, it's not very easy to manage my avatar since it is fetched from the related forum profile (if there is any) and this call slows down the back office.

Nice improvement, I agree on cutting off the call to the PrestaShop service to gain performance but I'm not in favor of creating a module for that. Could we remove the call and create a better built-in feature to handle avatars?

_Originally posted by @LouiseBonnard in https://github.com/PrestaShop/PrestaShop/pull/18346#issuecomment-611000014_

### Describe the solution you'd like:

Selecting an avatar is quite a basic feature so the idea here is to first offer users a simple way to define their avatar (points 1, 2, 6). Next, it is to make it flexible enough to have a great base for users willing to go further to be able to extend its usage or management (points 3, 4, 5). It includes:

1. :heavy_check_mark: **Cut off the call** to the PrestaShop service to gain performance, fixed by #18622, thanks @matks!

⚠️ do not forget to delete the information notification, cf. https://github.com/PrestaShop/PrestaShop/issues/18653#issuecomment-621243509

2.  :heavy_check_mark: #20114 **Add an image upload field** to the avatar in order to allow users to define their own profile picture - could be the same characteristics as the image upload fields of the Design > Theme & Logo page with the following allowed formats: .gif, .jpg, .png., and .svg

3. :heavy_check_mark:  #20207 **Create a hook** so that this space can also be used by modules or any other extensions requested by users.

4. :heavy_check_mark: #20207 **Manage its calls** in case several images use this hook: the last item using this hook will be the one displayed.

5. :heavy_check_mark: #20207 **Allow a Gravatar option field** for users who want to use it.

6. :heavy_check_mark: #20430 **Extend the avatar usage to profiles** and allow the SuperAdmin profile (only) to assign a default avatar to an employee category in the Team > Profiles section of the back office. 

:notebook: Avatar are displayed in Logs Page

### Alternatives you've considered

#18346

### PR DESCRIPTION
**Pull request by Progi1984**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Allow employee to choose their own avatar
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #18653
| How to test?  | * The current avatar of the current employee is the Preston Image.<br>* When you upload your avatar in the employee form, the avatar is the dropdown menu <br>image

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20114)

### CODE REVIEW

File: src/Adapter/Image/Uploader/EmployeeImageUploader.php [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+     * Deletes old image
+     *
+     * @param $id
+     */
+    private function deleteOldImage($id)
+    {
+        $employee = new Employee($id);
+        $employee->deleteImage();
+
+        $currentLogo = _PS_TMP_IMG_DIR_ . 'employee_mini_' . $id . '.jpg';
```
Comment (matks) [inline src/Adapter/Image/Uploader/EmployeeImageUploader.php]:
```suggestion
        $currentImage = _PS_TMP_IMG_DIR_ . 'employee_mini_' . $id . '.jpg';
```

File: src/Adapter/Image/Uploader/EmployeeImageUploader.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+        if (file_exists($currentLogo)) {
```
Comment (matks) [inline src/Adapter/Image/Uploader/EmployeeImageUploader.php]:
```suggestion
        if (file_exists($currentImage)) {
```

File: src/Adapter/Image/Uploader/EmployeeImageUploader.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            unlink($currentLogo);
```
Comment (matks) [inline src/Adapter/Image/Uploader/EmployeeImageUploader.php]:
```suggestion
            unlink($currentImage);
```

File: src/Adapter/Profile/Employee/QueryHandler/GetEmployeeForEditingHandler.php [hunk 4]
```diff
@@ -52,17 +68,40 @@ public function handle(GetEmployeeForEditing $query)
...
 
+        $avatarUrl = $this->getAvatarUrl($employeeId->getValue());
+
         return new EditableEmployee(
             $employeeId,
             new FirstName($employee->firstname),
             new LastName($employee->lastname),
             new Email($employee->email),
-            $employee->getImage(),
+            $avatarUrl ? $avatarUrl['path'] : $employee->getImage(),
```
Comment (matks) [inline src/Adapter/Profile/Employee/QueryHandler/GetEmployeeForEditingHandler.php:78]: I have some trouble understanding this line 😅 

If there is an avatar URL, we use the avatar URL. Else we use the employee image.

The avatar URL and the employee image are not the same things ? 😅

**Review by matks — DISMISSED** [review]
PR looks good to me.
I have
- 1 suggestion
- 1 question

But I approve the PR because both are not blocking

File: src/Adapter/Image/Uploader/EmployeeImageUploader.php (hunk 1 above)
Comment (Progi1984) [reply src/Adapter/Image/Uploader/EmployeeImageUploader.php]: Fixed

File: src/Adapter/Image/Uploader/EmployeeImageUploader.php (hunk 2 above)
Comment (Progi1984) [reply src/Adapter/Image/Uploader/EmployeeImageUploader.php]: Fixed

File: src/Adapter/Image/Uploader/EmployeeImageUploader.php (hunk 3 above)
Comment (Progi1984) [reply src/Adapter/Image/Uploader/EmployeeImageUploader.php]: Fixed

File: src/Adapter/Profile/Employee/QueryHandler/GetEmployeeForEditingHandler.php (hunk 4 above)
Comment (Progi1984) [reply src/Adapter/Profile/Employee/QueryHandler/GetEmployeeForEditingHandler.php:78]: If the avatar URL doesn't exist, the image employee returns the default image (aka Preston).

**Review by matks — DISMISSED** [review]

File: classes/Employee.php [hunk 5]
```diff
@@ -584,6 +584,13 @@ public function isSuperAdmin()
      */
     public function getImage()
     {
+        $imagePath = $this->image_dir . $this->id . '.jpg';
+        if (file_exists($imagePath) && is_file($imagePath)) {
```
Comment (PierreRambaud) [inline classes/Employee.php]:
```suggestion
        if (file_exists($imagePath)) {
```
You try to reach a jpg file, not sure it's needed

File: src/Adapter/Image/Uploader/EmployeeImageUploader.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+     * @param $id
+     */
+    private function deleteOldImage($id)
+    {
+        $employee = new Employee($id);
+        $employee->deleteImage();
+
+        $currentImage = _PS_TMP_IMG_DIR_ . 'employee_mini_' . $id . '.jpg';
+
+        if (file_exists($currentImage) && is_file($currentImage)) {
```
Comment (PierreRambaud) [inline src/Adapter/Image/Uploader/EmployeeImageUploader.php]:
```suggestion
        if (file_exists($currentImage)) {
```

File: src/Adapter/Image/Uploader/EmployeeImageUploader.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function upload($employeeId, UploadedFile $image)
+    {
+        $this->checkImageIsAllowedForUpload($image);
+        $tempImageName = $this->createTemporaryImage($image);
+        $this->deleteOldImage($employeeId);
+
+        $destination = _PS_EMPLOYEE_IMG_DIR_ . $employeeId . '.jpg';
```
Comment (PierreRambaud) [inline src/Adapter/Image/Uploader/EmployeeImageUploader.php]: `_PS_EMPLOYEE_IMG_DIR_` should be injected in the class no?

File: src/Core/Form/IdentifiableObject/DataHandler/EmployeeFormDataHandler.php [hunk 8]
```diff
@@ -118,6 +127,12 @@ public function create(array $data)
             isset($data['shop_association']) ? $data['shop_association'] : $this->defaultShopAssociation
         ));
 
+        /** @var UploadedFile $uploadedAvatar */
+        $uploadedAvatar = $data['avatarUrl'];
```
Comment (PierreRambaud) [inline src/Core/Form/IdentifiableObject/DataHandler/EmployeeFormDataHandler.php]: Is this key always exists?

**Review by PierreRambaud — COMMENTED** [review]
Few comments ;)

File: classes/Employee.php (hunk 5 above)
Comment (Progi1984) [reply classes/Employee.php]: Approved ;)

File: src/Adapter/Image/Uploader/EmployeeImageUploader.php (hunk 7 above)
Comment (Progi1984) [reply src/Adapter/Image/Uploader/EmployeeImageUploader.php]: Approved ;)

**Review by matks — APPROVED** [review]

File: src/Core/Form/IdentifiableObject/DataHandler/EmployeeFormDataHandler.php (hunk 8 above)
Comment (Progi1984) [reply src/Core/Form/IdentifiableObject/DataHandler/EmployeeFormDataHandler.php]: Fixed

File: src/Adapter/Image/Uploader/EmployeeImageUploader.php (hunk 6 above)
Comment (Progi1984) [reply src/Adapter/Image/Uploader/EmployeeImageUploader.php]: Fixed

**Review by PierreRambaud — APPROVED** [review]

**Comment by Progi1984** [comment]
Thanks @sarahdib

**Pull request merged**